File: apps/algo/src/main/java/sort/SelectionSort.java
package sort;

import static util.Arr.swap;

//2.2-2
public class SelectionSort implements IntSorter {

  @Override
  public int[] sort(int[] input) {
    for (int current = 0; current < input.length; current++) {
      int smallest = indexOfSmallest(input, current, input.length);
      swap(input, current, smallest);
    }
    return input;
  }

  private int indexOfSmallest(int[] input, int rangeFrom, int rangeTo) {
    int minIndex = rangeFrom;
    int min = input[minIndex];
    for (int i = rangeFrom; i < rangeTo; i++) {
      if (input[i] < min) {
        min = input[i];
        minIndex = i;
      }
    }
    return minIndex;
  }

}

File: books/algo/sort/selection_sort.go
package sort

//2.2-2
func SelectionSort(arr []int) {
	l := len(arr)
	if l == 0 || l == 1 {
		return
	}
	for i := 0; i < l-1; i++ {
		mini := i
		min := arr[mini]
		for j := i; j < l; j++ {
			if arr[j] < min {
				min = arr[j]
				mini = j
			}
		}
		arr[i], arr[mini] = arr[mini], arr[i]
	}
}
